src/ui/widgets/XYPlot/xyPlotOptions.ts

```diff
@@ -0,0 +1,259 @@
+      tmpY = tmpY.slice(0, numPoints);
+      tmpX = tmpX.slice(0, numPoints);
+    }
+
+    data.x = tmpX;
+    data.y = tmpY;
+    dataSet.push(data);
+  });
+  return dataSet;
+}
+
+/**
+ * Create plotly axis object for each axisOptions
+ * in the axes array.
+ * @param axesList list of all the axes information
+ * from the file
+ * @returns a Plotly formatted axis object
+ */
+export function createAxes(
+  axesList: Axis[],
+  font: React.CSSProperties
+): NewAxisSettings[] {
+  const formattedAxes: any[] = [];
+  let idx = 0;
+  for (const axis of axesList) {
+    const _newAxis: NewAxisSettings = {
+      autorange: true,
+      showline: true,
+      visible: axis.visible,
+      showgrid: axis.showGrid,
+      griddash: axis.dashGridLine,
+      gridwidth: 0.5,
+      gridcolor: axis.axisColor?.toString(),
+      tickcolor: axis.axisColor?.toString(),
+      zeroline: false,
+      automargin: true,
+      minor: {
+        ticks: "outside"
+      },
+      tickformat: axis.scaleFormat
+    };
+    // Only add title labels if they exist
+    if (axis.axisTitle !== "") {
+      _newAxis.title = {
+        text: axis.axisTitle,
+        standoff: 0
+      };
+      _newAxis.titlefont = {
+        family: font ? font.fontFamily : "Liberation sans, sans-serif",
+        size: font.fontSize ? font.fontSize : 12
+      };
+    }
+    _newAxis.range = [axis.minimum, axis.maximum];
+    _newAxis.autorange = false;
+
+    if (axis.logScale) _newAxis.type = "log";
+
+    // More than 2 axes, determine position and location of axes
+    if (idx > 1) {
+      _newAxis.side = axis.leftBottomSide ? "left" : "right";
+      _newAxis.overlaying = axis.yAxis ? "y" : "x";
+      _newAxis.position = 0;
+      _newAxis.anchor = "free";
+      // Only shift to add space for 2nd y-axis if it is overlapping and visible
+      if (axis.visible)
+        formattedAxes[0].domain = [0.08 * (axesList.length - 1), 1];
+    }
+    formattedAxes.push(_newAxis);
+    idx += 1;
+  }
+  return formattedAxes;
+}
+
+/**
+ * Calculates the axis range within the maximum
+ * and minimum limits using the highest and lowest
+ * data values. Defaults to range specified from file
+ * if calculated limits are larger.
+ * @param oldAxis options object read from file
+ * @param newAxis plotly axis options object
+ * @param dataSet array of PV data
+ * @param index the index of the axis
+ * @returns modified axis options object
+ */
+export function calculateAxisLimits(
+  oldAxis: Axis,
+  newAxis: NewAxisSettings,
+  dataSet: any[]
+): NewAxisSettings {
+  // If autoscale not enabled, do not rescale axes
+  if (!oldAxis.autoScale) return newAxis;
+  // Find axis limits
+  const [axMin, axMax]: [number, number] = newAxis.range;
+  // Determine if we are using x or y data
+  let dataType = "x";
+  if (oldAxis.index > 0 && oldAxis.yAxis) dataType = "y";
+  // Combine all data we need into an array of arrays
+  const allData = dataSet.map(val => val[dataType]);
```

Comment:
This might imply some copying which I would be wary of on performance grounds. There may be a more efficient way.

Reply:
This is where I don't know how TS handles references; in python it wouldn't copy but here I am not sure.

You could iterate and take the max of the maxes and the min of the mins of each array.

Reply:
I think `.map()` performs a shallow copy of the object. Iterating may indeed be quicker.